core/drivers/crypto/hisilicon/sec/sec_cipher.c

```diff
@@ -0,0 +1,773 @@
+	case TEE_ALG_SM4_ECB_NOPAD:
+	case TEE_ALG_SM4_CTR:
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static int crypto_set_calg(struct sec_cipher_ctx *c_ctx, const uint32_t alg)
+{
+	int ret = TEE_SUCCESS;
+
+	switch (alg) {
+	case TEE_MAIN_ALGO_DES:
+		c_ctx->alg = C_ALG_DES;
+		break;
+	case TEE_MAIN_ALGO_DES3:
+		c_ctx->alg = C_ALG_3DES;
+		break;
+	case TEE_MAIN_ALGO_AES:
+		c_ctx->alg = C_ALG_AES;
+		break;
+	case TEE_MAIN_ALGO_SM4:
+		c_ctx->alg = C_ALG_SM4;
+		break;
+	default:
+		EMSG("Invalid cipher type 0x%"PRIx8, alg);
+		ret = TEE_ERROR_NOT_IMPLEMENTED;
+		break;
+	}
+
+	return ret;
+}
+
+static int crypto_set_cmode(struct sec_cipher_ctx *c_ctx, const uint32_t mode)
+{
+	int ret = TEE_SUCCESS;
+
+	switch (mode) {
+	case TEE_CHAIN_MODE_ECB_NOPAD:
+		c_ctx->mode = C_MODE_ECB;
+		break;
+	case TEE_CHAIN_MODE_CBC_NOPAD:
+		c_ctx->mode = C_MODE_CBC;
+		break;
+	case TEE_CHAIN_MODE_XTS:
+		c_ctx->mode = C_MODE_XTS;
+		break;
+	case TEE_CHAIN_MODE_CTR:
+		c_ctx->mode = C_MODE_CTR;
+		break;
+	default:
+		EMSG("Invalid cipher mode type 0x%"PRIx32, mode);
+		ret = TEE_ERROR_NOT_IMPLEMENTED;
+		break;
+	}
+
+	return ret;
+}
+
+static TEE_Result sec_cipher_alloc_ctx(void **ctx, uint32_t algo)
+{
+	struct sec_cipher_ctx *c_ctx = NULL;
+	int ret = 0;
+
+	if (!ctx) {
+		EMSG("Ctx is NULL");
+		return TEE_ERROR_STORAGE_NO_SPACE;
+	}
+
+	ret = cipher_algo_check(algo);
+	if (ret != 0)
+		return ret;
+
+	c_ctx = (struct sec_cipher_ctx *)malloc(sizeof(struct sec_cipher_ctx));
+	if (!c_ctx) {
+		EMSG("c_ctx is NULL");
+		return TEE_ERROR_STORAGE_NO_SPACE;
+	}
+
+	ret = crypto_set_calg(c_ctx, TEE_ALG_GET_MAIN_ALG(algo));
+	if (ret)
+		goto free_c_ctx;
+
+	ret = crypto_set_cmode(c_ctx, TEE_ALG_GET_CHAIN_MODE(algo));
+	if (ret)
+		goto free_c_ctx;
+
+	c_ctx->qp = hisi_sec_create_qp(HISI_QM_CHANNEL_TYPE0);
+	if (!c_ctx->qp) {
+		ret = TEE_ERROR_BUSY;
+		goto free_c_ctx;
+	}
+
+	if (c_ctx->qp->qm->version == HISI_QM_HW_V2) {
+		c_ctx->qp->fill_sqe = sec_cipher_bd_fill;
+		c_ctx->qp->parse_sqe = sec_cipher_bd_parse;
+	} else {
+		c_ctx->qp->fill_sqe = sec_cipher_bd3_fill;
+		c_ctx->qp->parse_sqe = sec_cipher_bd3_parse;
+	}
+
+	c_ctx->offs = 0;
+	(*ctx) = c_ctx;
+
+	return TEE_SUCCESS;
+
+free_c_ctx:
+	free(c_ctx);
+
+	return ret;
+}
+
+static void sec_cipher_free_ctx(void *ctx)
+{
+	struct sec_cipher_ctx *c_ctx = (struct sec_cipher_ctx *)ctx;
+
+	if (!c_ctx)
+		return;
+
+	hisi_qm_release_qp(c_ctx->qp);
+	memset(c_ctx->key, 0, c_ctx->key_len);
+	free(c_ctx);
+}
+
+static TEE_Result sec_cipher_init(struct drvcrypt_cipher_init *dinit)
+{
+	struct sec_cipher_ctx *c_ctx = NULL;
+	TEE_Result ret = 0;
+
+	if (!dinit || !dinit->ctx || !dinit->key1.data) {
+		EMSG("drvcrypt_cipher init param error");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	c_ctx = (struct sec_cipher_ctx *)dinit->ctx;
+
+	ret = sec_cipher_set_key(c_ctx, dinit->key1.data, dinit->key1.length,
+				 dinit->key2.data, dinit->key2.length);
+	if (ret != 0)
```

Comment:
this ius why you should use `enum hisi_drv_status` as return value type for your function. The line above and line 580 below mess up between `enum hisi_drv_status` and `TEE_Result` types values. Using explicit `enum hisi_drv_status` return type instead of raw `int32` would have GCC to help you found and address such issues.

Another solution could be to use on `TEE_Result` type and code in your driver, assuming the available `TEE_Result` matches your driver expectation.